### PR TITLE
Add initial support for the AnkerMake M5C

### LINKED ISSUE #137: [BUG] M5C support - failed to decode mqtt message on current HEAD
**Describe the bug**
Cannot decode MQTT msessages with various payloads using commit bcf6855.

One of several messages that fail to decode:

[*] Trying connect to printer AnkerMake (EUPRAKM-012816-EJPZG) over pppp using ip 
[E] Failed to decode mqtt message
 Exception: Unsupported mqtt message format (expected 2, but found 1)
 Message : 0x4d 0x41 0x89 0x00 0x05 0x01 0x01 0x05 0x46 0xc0 0x01 0x00 0xac 0xf0 0x9e 0x50 0x04 0x0a 0x00 0x00 0x00 0x00 0x00 0x00 0x91 0x19 0x86 0xc8 0x5c 0x38 0x6b 0x3e 0x53 0x8d 0x8b 0xc2 0xec 0xb8 0xb2 0xbb 0x83 0xcb 0xad 0x14 0x78 0x10 0x38 0xef 0xcc 0xfa 0xa1 0xbd 0x1d 0x86 0x2c 0x9a 0x26 0xaf 0x45 0xbf 0xfb 0x59 0x84 0xc5 0x11 0xd4 0x5d 0x56 0xdc 0x99 0xcc 0x43 0xc5 0x28 0x06 0x0a 0xdd 0xe7 0xac 0x85 0x54 0xfb 0x10 0x87 0x64 0xa8 0xae 0x67 0x6c 0x01 0xce 0xf1 0x31 0xd5 0xcb 0x40 0x52 0xb3 0x30 0x78 0x74 0x1e 0x3a 0xc6 0x2c 0xf0 0x19 0x20 0x24 0x1d 0x94 0xbe 0x15 0xc4 0x7e 0x22 0x47 0x97 0x67 0x1e 0x7d 0x42 0x0a 0x9c 0x36 0xd2 0x17 0x97 0xc3 0x02 0xc5 0xa3 0xd9 0x3e 0x3b 0xc3 0x14

**To Reproduce**
1. Connect an M5C printer
2. Upgrade to latest firmware
3. Clone ankermake-m5-protocol
4. Start the ankerctl webserver
5. See error

**Expected behavior**
Expected MQTT messages to decode, MQTT/PPPP buttons to turn green and ability to upload gcode files

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS: Arch x64
 - Browser Firefox
- Firmware version: AnkerMake M5C firmware version V3.1.25

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by treitmayr**
This PR fixes #137 and allows to monitor MQTT messages of the AnkerMake M5C on the command line as well as observe the printer status via the Web GUI. It is also possible to send print jobs from your slicer to the printer without using AnkerMake Slicer or Studio (see README.md). The M5's additional hardware is still expected to exist, therefore the Web GUI will still pretend to wait for a camera connection, for example. This will have to be fixed in the future.

Note that I was only able to test with my M5C but not with an M5 for unwanted side-effects.

Eventually, if this patch is confirmed by others to work fine, we should add some related information to the README file.

### CODE REVIEW

**Comment by treitmayr** [comment]
I just realized that sending g-code from the command line results in an error - will investigate further.

**Comment by treitmayr** [comment]
Found the issue! Apparently the `padding` field was not being set in `make_mqtt_pkt()`. Changed it to an empty binary string to let `.pack()` fill it up depending on the `m5` byte. Note that the `m5` byte will still be `2` for sent messages, not `1` as found in received messages for the M5C. But that does not seem to hurt, as I was able to home my printer's head by sending "G28;" from the command line.

**Comment by billyjbryant** [comment]
fixes #137 #147

**Comment by treitmayr** [comment]
One more observation - the PPPP badge in the web interface will currently not turn green for the M5C because, after reading the code, this is actually the status of the video connection. So the PPPP connection may be established correctly after all and sending gcode files will work just fine.

**Comment by billyjbryant** [comment]
Maybe we should add a check for the badge that does a health check on PPPP for M5C to turn the status green in absence of the video feed

**Comment by treitmayr** [comment]
> Maybe we should add a check for the badge that does a health check on PPPP for M5C to turn the status green in absence of the video feed

I may take a look into that as time permits. However, as it supposedly does not affect the overall functionality I assume this can be done via a separate PR.

**Comment by treitmayr** [comment]
Any review comments for this PR? Are there things to change before it can be merged?

**Comment by treitmayr** [comment]
As there was no feedback yet to the previous commits anyway, I ended up adding the feature suggested by @billyjbryant. I.e. the PPPP status is now correctly displayed for both M5 and M5C by expecting a PPPP message at least every 3 seconds. This makes the badge turn green, otherwise it is displayed as orange. The 3 seconds are a valid assumption IMO because the printer sends at least an AlivePkt every second.
A new VIDEO badge takes over the original PPPP badge.

This latest commit also adds conditions around parts of the web interface which are not applicable to the M5C. So M5C users will not see the rotating lines in the video area anymore, nor the controls for the camera or the VIDEO badge.

I have tested this with my M5C (which "luckily" drops the Wifi connection very often when connected to my Unifi AP), but tests with an M5 would be appreciated to see if things work the same as before (except for the added VIDEO badge).

**Comment by treitmayr** [comment]
I opened this PR almost one month ago, and while I admittedly added more changes later on, there was no comment to the specific code changes or prospects for when this can be merged from any of the maintainers of this great piece of software. In the meantime I direct people with various issues to my own [integration branch](https://github.com/treitmayr/ankermake-m5-protocol/tree/m5c-integration), but this cannot be a permanent situation IMO.

Any thoughts about the changes and when this can be merged? Or is this project not actively maintained and developed further anymore?

**Comment by bakkerv** [comment]
Deployed this branch this morning. Works like a charm, way better than manually exporting gcode and opening Ankermake Studio!

**Review by Suppe12 — APPROVED** [review]
Code gets the Ankermake M5C perfectly working, got it running on my own setup

**Review by TheeStickmahn — COMMENTED** [review]
very good very work

still haven't tested sending prints over from ps so no approval yet
but very good job

**Review by TheeStickmahn — COMMENTED** [review]
Uploading files from prusaslicer gets stuck on "RESOLVING ADDRESS".

Uploading files: no worky :(
Everything else: yes worky :)

**Comment by treitmayr** [comment]
> Uploading files from prusaslicer gets stuck on "RESOLVING ADDRESS".

This is very likely an issue of PrusaSlicer and not of this PR or ancertctl in general, see https://github.com/prusa3d/PrusaSlicer/issues/10944 . I saw the same issue a while ago but the latest versions of PS seem to have fixed it.

**Comment by TheeStickmahn** [comment]
> This is very likely an issue of PrusaSlicer and not of this PR or ancertctl in general, see [prusa3d/PrusaSlicer#10944](https://github.com/prusa3d/PrusaSlicer/issues/10944) . I saw the same issue a while ago but the latest versions of PS seem to have fixed it.

I'm running PS 2.8, so it isn't that.
Turns out the printer claims to have connected to wifi but doesn't actually connect.
They should really add a network test on the printer.

tldr: fck you ankermake

**Comment by ErebusBat** [comment]
Was this work moved somewhere?

**Comment by treitmayr** [comment]
This and other PRs are now incorporated in https://github.com/anselor/ankermake-m5-protocol/tree/exiles-1.1 . Unfortunately, I have to admit that the documentation is still not up to date and matching the now recommended way (which is a simple login and printer search via the web interface).
